app/src/main/java/org/openobservatory/ooniprobe/activity/RunTestActivity.java

```diff
@@ -0,0 +1,236 @@
+                    }
+                }
+                else {
+                    Alert.alertDialogTwoButtons(this, getString(R.string.ooniprobe_outdate), getString(R.string.ooniprobe_outdate_msg), new DialogInterface.OnClickListener() {
+                        public void onClick(DialogInterface dialog, int id) {
+                            final String appPackageName = getPackageName(); // getPackageName() from Context or Activity object
+                            try {
+                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + appPackageName)));
+                            } catch (android.content.ActivityNotFoundException anfe) {
+                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + appPackageName)));
+                            }
+                        }
+                    });
+                }
+            }
+        }
+    }
+
+    public void configureScreen(String td, String ta){
+        TestData.getInstance(this, this).addObserver(this);
+         urls = (TextView) findViewById(R.id.urls);
+
+        title = (TextView) findViewById(R.id.run_test_message);
+        if (td != null)
+            title.setText(td);
+        else
+            title.setText(getString(R.string.run_test_message));
+
+        String test = NetworkMeasurement.getTestName(this, test_name);
+        testTitle = (TextView) findViewById(R.id.test_title);
+        testTitle.setText(test);
+
+        testImage = (ImageView) findViewById(R.id.test_logo);
+        testImage.setImageResource(NetworkMeasurement.getTestImageBig(test_name));
+
+        ArrayList<String> listItems = new ArrayList<>();
+        final ArrayList<String> urlItems = new ArrayList<>();
+        if (ta != null){
+            try {
+                JSONObject taJson = new JSONObject(ta);
+                JSONArray urls = taJson.getJSONArray("urls");
+                for (int i = 0; i < urls.length(); i++)
+                    urlItems.add(urls.getString(i));
+            } catch (JSONException e) {
+                e.printStackTrace();
+            }
+            listItems.addAll(urlItems);
+        }
+
+        if (listItems.size() == 0 && test_name.equals("web_connectivity")){
+            listItems.add(getString(R.string.random_sampling_urls));
+            mUrlListAdapter = new UrlListAdapter(this, listItems, false);
+        }
+        else
+            mUrlListAdapter = new UrlListAdapter(this, listItems, true);
+
+        if (listItems.size() == 0)
+            urls.setVisibility(View.INVISIBLE);
+
+        runButton = (AppCompatButton) findViewById(R.id.run_test_button);
+        runButton.setOnClickListener(
+                new ImageButton.OnClickListener() {
+                    public void onClick(View v) {
+                        TestData.doNetworkMeasurements(getApplicationContext(), test_name, urlItems);
```

Comment:
I am a bit confused by the fact that sometimes you call `TestData`'s singleton and some other times you don't. This makes it more complex for me to reason about the lifecycle of the singleton.

Reply:
TestData.doNetworkMeasurements is always called in the same way. Because it's already been created

Reply:
What I mean is that sometimes you access the singleton of `TestData` (for example for setting the observers) and in other cases like this instead you call a static method. This does not make obvious what is the singleton used for and whether the singleton is used or not when you call the static method, unless one reads the code.

Reply:
So I guess I should change everything to singleton